[kernel 3c/n] Decouple validation cache initialization from `ArgsManager`

dongcarl opened the pull request:
This is part of the `libbitcoinkernel` project: #24303, https://github.com/bitcoin/bitcoin/projects/18

This PR is **_NOT_** dependent on any other PRs.

-----

a.k.a. "Stop calling `gArgs.GetIntArg("-maxsigcachesize")` from validation code"

This PR introduces the `ValidationCacheSizes` struct and its corresponding `ApplyArgsManOptions` function, removing the need to call `gArgs` from `Init{Signature,ScriptExecution}Cache()`. This serves to further decouple `ArgsManager` from `libbitcoinkernel` code.

More context can be gleaned from the commit messages.

dongcarl committed:
tests: Reduce calls to InitS*Cache()

In src/test/fuzz/script_sigcache.cpp, we should really be setting up a
full working BasicTestingSetup. The initialize_ function is only run
once anyway.

In src/test/txvalidationcache_tests.cpp, the Dersig100Setup inherits
from BasicTestingSetup, which should have already set up a global script
execution cache without the need to explicitly call
InitScriptExecutionCache.

dongcarl committed:
cuckoocache: Return approximate memory size

Returning the approximate total size eliminates the need for
InitS*Cache() to do nElems*sizeof(uint256). The cuckoocache has a better
idea of this information.

dongcarl committed:
validationcaches: Add and use ValidationCacheSizes

Also:

- Make DEFAULT_MAX_SIG_CACHE_SIZE into constexpr
  DEFAULT_MAX_SIG_CACHE_BYTES to utilize the compile-time integer
  arithmetic overflow checking available to constexpr.
- Fix comment (MiB instead of MB) for DEFAULT_MAX_SIG_CACHE_BYTES.
- Pass in max_size_bytes parameter to InitS*Cache(), modify log line to
  no longer allude to maxsigcachesize being split evenly between the two
  validation caches.
- Fix possible integer truncation and add a comment.

[META] I've kept the integer types as int64_t in order to not introduce
       unintended behaviour changes, in the next commit we will make
       them size_t.

dongcarl committed:
validationcaches: Use size_t for sizes

...also move the 0-clamping logic to ApplyArgsManOptions, where it
   belongs.

dongcarl commented:
So it seems like the static asserts I added revealed that on `MAX_MAX_SIG_CACHE_SIZE * (1 << 20)` overflows on 32-bit systems... This is an existing bug.

https://cirrus-ci.com/task/5077719523262464?logs=ci#L2274

dongcarl commented:
Push 3168fbf6ae9493df25727a11925a2f6677f30af2 -> 7f443a019ef90ee1a2a324f0b6a9abc98ef92ce5
- Abolish the arbitrary `MAX_MAX_SIG_CACHE_SIZE` limit and instead return a friendly error when unable to allocate memory for validation caches
- Make `DEFAULT_MAX_SIG_CACHE_SIZE` into `constexpr DEFAULT_MAX_SIG_CACHE_SIZE_BYTES` to utilize the compile-time integer arithmetic overflow checking available to `constexpr`s.

dongcarl committed:
cuckoocache: Check for uint32 overflow in setup_bytes

This fixes an potential overflow which existed prior to this patchset.

If CuckooCache::cache<Element, Hash>::setup_bytes is called with a
`size_t bytes` which, when divided by sizeof(Element), does not fit into
an uint32_t, the implicit conversion to uint32_t in the call to setup
will result in an overflow.

At least on x86_64, this overflow is possible:

static_assert(std::numeric_limits<size_t>::max() / 32 <= std::numeric_limits<uint32_t>::max());
static_assert(std::numeric_limits<size_t>::max() / 4 <= std::numeric_limits<uint32_t>::max());

This commit detects such cases and signals to callers that the `size_t
bytes` input is too large.

dongcarl commented:
Push 7f443a019ef90ee1a2a324f0b6a9abc98ef92ce5 -> 1bcae96af276b8f4ea4ea15f9f3a5c8e2180c44c
- Add overflow check when converting from total size to number of elements in `setup_bytes`

dongcarl commented:
Pushed 1bcae96af276b8f4ea4ea15f9f3a5c8e2180c44c -> 8e407571e99b0d4aad7f225b8197f35c6b6cea45
- Rebased over master

ryanofsky commented on src/script/sigcache.cpp:
```diff
@@ -92,14 +92,19 @@ static CSignatureCache signatureCache;
-    size_t nElems = signatureCache.setup_bytes(nMaxCacheSize);
-    LogPrintf("Using %zu MiB out of %zu/2 requested for signature cache, able to store %zu elements\n",
-            (nElems*sizeof(uint256)) >>20, (nMaxCacheSize*2)>>20, nElems);
+    size_t nMaxCacheSize = std::max((int64_t)0, gArgs.GetIntArg("-maxsigcachesize", DEFAULT_MAX_SIG_CACHE_SIZE) / 2) * ((size_t) 1 << 20);
```
In commit "cuckoocache: Return approximate memory size, failure" (6d0d7d9759c90c0533528893ce084f04be9387ee)

I think it is worth adding a release note in this commit documenting that `-maxsigcachesize` internal cap is removed, so existing configurations with larger values may now be broken or demand more memory.

dongcarl replied:
Fixed in 0940f47e5e, let me know if the release note message is missing anything or feel free to resolve this!

ryanofsky commented on src/cuckoocache.h:
```diff
@@ -367,11 +368,17 @@ class cache
     {
-        auto num_elems = setup(bytes/sizeof(Element));
+        size_t requested_num_elems = bytes / sizeof(Element);
+        if (std::numeric_limits<uint32_t>::max() < requested_num_elems) {
```
In commit "cuckoocache: Check for uint32 overflow in setup_bytes" (9f9911595f90b920fca71e53d2807f5b6feca137)

Adding this check seems good but what was motivation? Is it just a problem with the previous code that you noticed and decided to fix? Or does it help with something that comes later? Would be good to say in commit message how this change relates to the PR

dongcarl replied:
Added some more context in eb8b6a9c5e, feel free to reopen if not clear!

ryanofsky replied:
> Added some more context in [eb8b6a9](https://github.com/bitcoin/bitcoin/commit/eb8b6a9c5ea9d4aea8d542adb219d8d7b15ddce0), feel free to reopen if not clear!

CuckooCahce is misspelled but thanks for the update this is clear!

ryanofsky commented on src/init.cpp:
```diff
@@ -103,20 +105,22 @@
 #endif
 
+using kernel::ValidationCacheSizes;
+using node::ApplyArgsManOptions;
 using node::CacheSizes;
 using node::CalculateCacheSizes;
-using node::ChainstateLoadVerifyError;
 using node::ChainstateLoadingError;
+using node::ChainstateLoadVerifyError;
```
In commit "validationcaches: Add and use ValidationCacheSizes" (bedef3249cfa013a48347b19fb58712a8d0fdd28)

I guess all these changes come from your editor doing case-insensitive sorts, which maybe is a preference. When I added these statements here and other places in #23497 I did purposefully use case-sensitive sorts, since I think case sensitive sorts make more sense in a case sensitive language, and look less glitchy visually, and do a better job grouping related identifiers together. You may perhaps want to keep this sorting, if it is not too much of a pain.

dongcarl replied:
I agree! Pesky editor settings... Fixed in caaa5be4ec

ryanofsky commented on src/node/validation_cache_args.cpp:
```diff
@@ -18,9 +21,10 @@ namespace node {
 void ApplyArgsManOptions(const ArgsManager& argsman, ValidationCacheSizes& cache_sizes)
 {
     if (auto max_size = argsman.GetIntArg("-maxsigcachesize")) {
+        size_t clamped_size_each = std::max<int64_t>(*max_size, 0) / 2 * (1 << 20);
```
In commit "validationcaches: Use size_t for sizes" (8e407571e99b0d4aad7f225b8197f35c6b6cea45)

Deleted comment about if size "is set to zero, setup_bytes creates the minimum possible cache (2 elements)", may be useful to keep to explain why 0 is the floor

ryanofsky reviewed (approved):
Code review ACK 8e407571e99b0d4aad7f225b8197f35c6b6cea45. Makes sense and cleans up some things nicely

dongcarl commented:
Push 8e407571e99b0d4aad7f225b8197f35c6b6cea45 -> 7dfe0c29ec5a11fe60030413660155edcadc3268
- Addressed Ryan's review comments (Thanks Ryan!)

dongcarl commented on src/bitcoin-chainstate.cpp:
```diff
@@ -61,8 +61,8 @@ int main(int argc, char* argv[])
     // performing the check with the signature cache.
-    InitSignatureCache();
-    InitScriptExecutionCache();
+    assert(InitSignatureCache());
```
Fixed in 0940f47e5e

ryanofsky reviewed (approved):
Code review ACK 7dfe0c29ec5a11fe60030413660155edcadc3268, just suggested changes since last review

dongcarl commented:
Push 7dfe0c29ec5a11fe60030413660155edcadc3268 -> 66767ab8395f940793c00371e2a14174ad46d6c8
- Fixed misspelling: https://github.com/bitcoin/bitcoin/pull/25527#discussion_r919337727

ryanofsky reviewed (approved):
Code review ACK 66767ab8395f940793c00371e2a14174ad46d6c8. Just cuckoocache spelling fix since last review.

I am the lonely reviewer here so would be good to have more review. This PR adds the first ever code to catch bad_alloc, and no one knows when the bad_alloc will strike. This is exciting stuff!

dongcarl commented:
Push 66767ab839 -> 365a04e9ac
- Addressed:
  - https://github.com/bitcoin/bitcoin/pull/25527#discussion_r921254010
  - https://github.com/bitcoin/bitcoin/pull/25527#discussion_r921262075

Thanks Ryan!

dongcarl commented on src/test/util/setup_common.cpp:
```diff
@@ -136,8 +136,8 @@ BasicTestingSetup::BasicTestingSetup(const std::string& chainName, const std::ve
-    InitSignatureCache();
-    InitScriptExecutionCache();
+    assert(InitSignatureCache());
+    assert(InitScriptExecutionCache());
```
Fixed in 51c08032b8

dongcarl commented on src/script/sigcache.cpp:
```diff
@@ -92,11 +92,11 @@ static CSignatureCache signatureCache;
 // To be called once in AppInitMain/BasicTestingSetup to initialize the
 // signatureCache.
-bool InitSignatureCache()
+bool InitSignatureCache(int64_t max_size)
```
Fixed in 1c89fb7f24

ryanofsky reviewed (approved):
Code review ACK 95805440101820cc246bb4379d8505d81074eb99. Since last review, just some changes assert -> Assert, max_size -> max_size_bytes and switching to designated initializers

ryanofsky reviewed (approved):
Code review ACK 915c8ff123c770f0cce82358256ab02b93018767. Just rebased since last review.

Would encourage more review of this PR. If it is not clear what this is doing I'd summarize it as "Stop calling gArgs.GetIntArg("-maxsigcachesize") from validation code"

aureleoules commented on src/cuckoocache.h:
```diff
@@ -326,19 +329,25 @@ class cache
+        try {
+            table.reserve(size);
+            table.resize(size);
+        } catch (std::bad_alloc const&) {
+            return std::nullopt;
+        }
```
This is never reached because there is already a global handler to handle bad allocs, so I think you can remove the `std::optional`.
https://github.com/bitcoin/bitcoin/blob/915c8ff123c770f0cce82358256ab02b93018767/src/init.cpp#L753-L764

dongcarl replied:
Oooof that sucks... There's no way to have a new handler that can access the `std::bad_alloc` object?

aureleoules replied:
AFAIK there doesn't seem to be a way :/

dongcarl replied:
Removed as of f66cd5b3aa, I think the existence of a new_handler makes this not too big of a problem now, and as was noted `-maxsigcachesize` is `DEBUG_ONLY`

dongcarl replied:
Closing, feel free to open if still a problem.

mzumsande commented on doc/release-notes-25527.md:
```diff
@@ -0,0 +1,8 @@
+Updated settings
+----------------
+
+- Specifying a `-maxsigcachesize` MiB value which is too large will now result
```
`-maxsigcachesize` is DEBUG_ONLY, should we cover these in release notes too?

dongcarl replied:
Very good point! Fixed in f66cd5b3aa

mzumsande reviewed (commented):
Concept ACK

dongcarl committed:
validationcaches: Abolish arbitrary limit

1. -maxsigcachesize is a DEBUG_ONLY option

2. Almost 7 years has passed since its semantics change in
   830e3f3d027ba5c8121eed0f6a9ce99961352572 from "number of entries" to
   "number of mebibytes"

3. A std::new_handler was added to the codebase after the original PR
   which introduced this limit, which will terminate immediately instead
   of causing trouble by being caught somewhere unexpected.

dongcarl commented:
Push 915c8ff123c770f0cce82358256ab02b93018767 -> d29f2fd9b49ac00a3721af7260dbbf59e9e8387c
- Realize that `-maxsigcachesize` is `DEBUG_ONLY`, remove release notes and adjust estimated importance of back-compatibility
  - Fixes https://github.com/bitcoin/bitcoin/pull/25527#discussion_r928336313
- Restore `CuckooCache::cache::setup` to return a non-optional again, since we can't actually catch a `std::bad_alloc` (see: https://github.com/bitcoin/bitcoin/pull/25527#discussion_r926395818)
- Fix log line for memory reporting, fixing:
  - https://github.com/bitcoin/bitcoin/pull/25527#discussion_r928304305
  - https://github.com/bitcoin/bitcoin/pull/25527#discussion_r928304361)
- Split out limit abolition to its own commit

dongcarl commented on src/script/sigcache.cpp:
```diff
@@ -92,14 +92,19 @@ static CSignatureCache signatureCache;
+
+    const auto [num_elems, approx_size_bytes] = *setup_results;
+    LogPrintf("Using %zu MiB out of %zu requested for signature cache, able to store %zu elements\n",
+              approx_size_bytes, nMaxCacheSize, num_elems);
```
The bytes thing should be fixed now in f66cd5b3aa

Since we now pass in the actual amount in e4d5bab88e (not amount * 2), I think we shouldn't say "/2". Since `libbitcoinkernel` users might specify different sizes, or we might decide to split it differently in the future.

ryanofsky commented on src/init.cpp:
```diff
@@ -550,7 +554,7 @@ void SetupServerArgs(ArgsManager& argsman)
     argsman.AddArg("-capturemessages", "Capture all P2P messages to disk", ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
     argsman.AddArg("-mocktime=<n>", "Replace actual time with " + UNIX_EPOCH_TIME + " (default: 0)", ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
-    argsman.AddArg("-maxsigcachesize=<n>", strprintf("Limit sum of signature cache and script execution cache sizes to <n> MiB (default: %u)", DEFAULT_MAX_SIG_CACHE_SIZE), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
+    argsman.AddArg("-maxsigcachesize=<n>", strprintf("Limit sum of signature cache and script execution cache sizes to <n> MiB (default: %u)", DEFAULT_MAX_SIG_CACHE_SIZE_BYTES >> 20), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
```
In commit "validationcaches: Add and use ValidationCacheSizes" (e4d5bab88e9acad53cb8d636d836d2cdf9bb49b8)

Throughout this commit could replace _size_bytes with _bytes. A lot of these names are pretty long and a bytes value is obviously a size.

dongcarl replied:
Fixed as of eeb224e4d3

ryanofsky commented on src/node/validation_cache_args.cpp:
```diff
@@ -0,0 +1,26 @@
+{
+    if (auto max_size = argsman.GetIntArg("-maxsigcachesize")) {
+        cache_sizes = {
+            .signature_cache_bytes = *max_size / 2 * (1 << 20),
```
In commit "validationcaches: Add and use ValidationCacheSizes" (e4d5bab88e9acad53cb8d636d836d2cdf9bb49b8)

I guess this would be a change in behavior, but maybe tweak order of operations so if a 1 MiB cache is requested this uses .5 MiB for each instead of 0

dongcarl replied:
Fixed as of eeb224e4d3, great catch!

ryanofsky reviewed (approved):
Code review ACK d29f2fd9b49ac00a3721af7260dbbf59e9e8387c

dongcarl commented:
Push d29f2fd9b49ac00a3721af7260dbbf59e9e8387c -> c7138d317e8db5874c01cbd44a64942200214d1a
- Address:
  - https://github.com/bitcoin/bitcoin/pull/25527#discussion_r933641007
  - https://github.com/bitcoin/bitcoin/pull/25527#discussion_r933641616

ryanofsky reviewed (commented):
Code review ACK c7138d317e8db5874c01cbd44a64942200214d1a, now rounding -maxsigcachesize down to nearest 1MiB instead of 2MiB (slight change in behavior) and renaming _size_bytes -> _bytes as suggested

glozow reviewed (commented):
code review ACK c7138d317e8db5874c01cbd44a64942200214d1a

happy to reACK after rebase

dongcarl commented:
Push c7138d317e -> 0f3a2532c3
- Rebased over master to resolve conflict
- Address: https://github.com/bitcoin/bitcoin/pull/25527#discussion_r936455075

dongcarl commented on src/script/sigcache.h:
```diff
@@ -10,12 +10,13 @@
 // systems). Due to how we count cache size, actual memory usage is slightly
 // more (~32.25 MB)
-static const unsigned int DEFAULT_MAX_SIG_CACHE_SIZE = 32;
+static constexpr size_t DEFAULT_MAX_SIG_CACHE_BYTES{32 << 20};
```
Fixed in 41c5201a90, thanks!

hernanmarino reviewed (approved):
Code review ACK .

glozow commented:
re ACK 0f3a2532c3

theStack reviewed (approved):
Code-review ACK 0f3a2532c38074dd9789d1c4c667db6ca46ff0ab

For a better understanding, I can (as always) highly recommend reading [this PR's review club notes (and chat log)](https://bitcoincore.reviews/25527) to other potential reviewers.

ryanofsky reviewed (approved):
Code review ACK 0f3a2532c38074dd9789d1c4c667db6ca46ff0ab. Rebase and comment tweak since last

ryanofsky commented:
I just realized I asked to rename `size` -> `size_bytes`, and then later asked rename `size_bytes` to `bytes` :facepalm:

At least it didn't go in a circle, though
